packages/myst-cli/src/transforms/execute.ts

```diff
@@ -0,0 +1,212 @@
+  };
+}
+
+/**
+ * Execute a code string in the given kernel, returning any outputs.
+ *
+ * @param kernel connection to an active kernel
+ * @param code code to execute
+ */
+async function executeCode(kernel: Kernel.IKernelConnection, code: string) {
+  const future = kernel.requestExecute({
+    code: code,
+  });
+
+  const outputs: IOutput[] = [];
+  future.onIOPub = (msg) => {
+    // Only listen for replies to this execution
+    if (
+      (msg.parent_header as any).msg_id !== future.msg.header.msg_id ||
+      (msg.parent_header as any).msg_type !== 'execute_request'
+    ) {
+      console.debug('Ignoring IOPub reply');
+      return;
+    }
+    switch (msg.header.msg_type) {
+      case 'stream':
+      case 'execute_result':
+      case 'error':
+      case 'display_data':
+        outputs.push(IOPubAsOutput(msg));
+        break;
+      default:
+        console.debug('Ignoring IOPub reply (unexpected message type)');
+        return;
+    }
+  };
+  let status: 'abort' | 'error' | 'ok' | undefined;
+  future.onReply = (msg: KernelMessage.IExecuteReplyMsg) => {
+    status = msg.content.status;
+  };
+  await future.done;
+  assert(status !== undefined);
+  return { status, outputs };
+}
+
+/**
+ * Evaluate an expression string in the given kernel, returning the singular result.
+ *
+ * @param kernel connection to an active kernel
+ * @param code code to execute
+ */
+async function evaluateExpression(kernel: Kernel.IKernelConnection, expr: string) {
+  // TODO: batch user expressions by cell?
+  const future = kernel.requestExecute({
+    code: '',
+    user_expressions: {
+      expr: expr,
+    },
+  });
+  let result: IExpressionResult | undefined;
+  future.onReply = (msg: KernelMessage.IExecuteReplyMsg) => {
+    switch (msg.content.status) {
+      case 'ok':
+        // Pull out expr
+        result = (msg.content.user_expressions as unknown as Record<string, IExpressionResult>)[
+          'expr'
+        ];
+        break;
+      case 'error':
+        assert(false); // We shouldn't hit this, because we don't evaluate expressions AND code simultaneously.
+        break;
+      default:
+        break;
+    }
+  };
+  await future.done;
+  // It doesn't make sense for this to be undefined instead of an error
+  assert(result !== undefined);
+  return { status: result.status, result };
+}
+
+type CacheKey = any;
+type CacheItem = (Code | InlineExpression)[];
+
+function buildCacheKey(mdast: (ICellBlock | InlineExpression)[]): any {
+  return undefined;
+}
+
+function getCache(key: CacheKey): CacheItem | undefined {
+  return undefined;
+}
+
+function setCache(key: CacheKey, value: CacheItem) {}
+
+type ICellBlockOutput = GenericNode & {
+  data: IOutput[];
+};
+
+type ICellBlock = GenericNode & {
+  children: (Code | ICellBlockOutput)[];
+};
+
+function isCellBlock(node: GenericNode): node is ICellBlock {
+  return node.type === 'block' && select('code', node) !== null && select('output', node) !== null;
+}
+
```

Comment:
Is there a better way to type this?

Reply:
Things look good?!

Reply:
@fwkoch did we have a `kind` on the block?

Reply:
We are tagging executable code blocks with `data.type = 'notebook-code'` - From notebooks - https://github.com/executablebooks/mystmd/blob/main/packages/myst-cli/src/process/notebook.ts#L24-L27 and from code cell directives - https://github.com/executablebooks/mystmd/blob/main/packages/myst-directives/src/code.ts#L215

However, I'm not sure we are relying on this field anywhere? In the myst-theme execution code at least, we find executable blocks in a similar way to your implementation: https://github.com/executablebooks/myst-theme/blob/855f29ec42a77deaa145481bfcab25b2d3c53757/packages/jupyter/src/execute/utils.ts#L8 The only difference is that function supports executable figures, i.e. `container` nodes with code/output children, in addition to blocks.